Source/WebCore/editing/Editing.cpp

```diff
@@ -1222,4 +1223,117 @@ HashSet<RefPtr<HTMLImageElement>> visibleImageElementsInRangeWithNonLoadedImages
     return result;
 }
 
+enum class RangeEndpoint : bool { Start, End };
+static std::optional<BoundaryPoint> visuallyContiguousBoundary(InlineBoxAndOffset&& fromPosition, InlineBoxAndOffset&& toPosition, TextDirection lineDirection, RangeEndpoint startOrEnd)
+{
+    struct VisuallyAdjacentTextRange {
+        Ref<Text> text;
+        unsigned startOffset { 0 };
+        unsigned endOffset { 0 };
+        unsigned minOffset { 0 };
+        unsigned maxOffset { 0 };
+        TextDirection direction { TextDirection::LTR };
+
+        unsigned totalLength() const { return maxOffset > minOffset ? maxOffset - minOffset : 0; }
```

Comment:
Maybe `std::max(maxOffset - minOffset, 0)`?

Reply:
oh no, unsigned arithmetic 😞

Reply:
Wired: `unsigned totalLength() const { return std::max(maxOffset, minOffset) - minOffset; }`

Reply:
SGTM, I'll go with the last approach (for `totalLength` and `visuallySelectedLength`)